### PR TITLE
add ingester interface

### PR DESCRIPTION
**Pull request by replay**
I would like to extend the Ingester by wrapping it in another struct, which overrides some methods and does additional work before calling the original Ingester methods.

Currently the API method `.RegisterIngester()` expects the concrete type `*ingester.Ingester`, so I can't pass it the extended ingester type. Outside the `api` package I also can't easily re-implement the `.RegisterIngester()` method, because it uses an unexported property of the API (`sourceIPs`). 
By changing the API method `RegisterIngester()` to expect an interface I'll be able to pass it the extended Ingester type and register its methods on the HTTP and the GRPC server.

What do you think about this approach?
An alternative to creating an Ingester interface would be to export that unexported property `.sourceIPs` which is used by `API.RegisterIngester()`, that way I could basically just re-implement what this method is doing outside the package, but that seems less elegant and might be more error-prone because if `RegisterIngester()` changes in the future I'll have to ensure that my copy stays up2date.

### CODE REVIEW

**Comment by pstibrany** [comment]
I would suggest moving new `Ingester` interface into `api` package (where it is actually used) while keeping `Ingester` type in `pkg/ingester` unchanged (no rename).

**Comment by replay** [comment]
> I would suggest moving new Ingester interface into api package (where it is actually used) while keeping Ingester type in pkg/ingester unchanged (no rename).

Thx, that's a great idea, I'll do that

File: pkg/api/api.go [hunk 1]
```diff
@@ -201,8 +200,15 @@ func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distrib
 	a.RegisterRoute("/ha-tracker", d.HATracker, false, "GET")
 }
 
+type ingester interface {
```
Comment (pracucci) [inline pkg/api/api.go:203]: Shouldn't be exported?

Comment (replay) [reply pkg/api/api.go:203]: For what I'm trying to do it's not necessary to export it. My goal is just to pass `RegisterIngesters()` an alternative struct which satisfies the same interface as the "original" ingester. Or is there another reason to export it?

Comment (pstibrany) [reply pkg/api/api.go:203]: I don't think export is necessary, but a comment explaining why we do this would be nice. Otherwise it looks like magic.

**Review by pstibrany — APPROVED** [review]
LGTM

File: pkg/api/api.go (hunk 1 above)
Comment (replay) [reply pkg/api/api.go:203]: ok, good point, I'll add one

**Review by pracucci — APPROVED** [review]

**Review by jtlisi — APPROVED** [review]
LGTM

**Pull request merged**